File: Tetris3D.java
// Tetris3D.java — трёхмерный тетрис на Java

import java.util.*;
import java.io.*;

public class Tetris3D {
    private static final int W=4, H=4, D=4;
    private int[][][] field = new int[W][H][D];
    private int score = 0, level = 1;
    private double fallInterval = 1.0;
    private List<int[]> currentPiece, nextPiece;
    private int[] currentPos = new int[3]; // x,y,z
    private boolean gameOver = false;
    private Random rand = new Random();
    private List<List<int[]>> shapes = new ArrayList<>();

    public Tetris3D() {
        // инициализация фигур
        int[][][] shapeData = {
            {{0,0,0},{1,0,0},{0,1,0},{1,1,0}},
            {{0,0,0},{1,0,0},{2,0,0},{3,0,0}},
            {{0,0,0},{1,0,0},{2,0,0},{1,1,0}},
            {{0,0,0},{1,0,0},{2,0,0},{0,1,0}},
            {{0,0,0},{1,0,0},{2,0,0},{2,1,0}},
            {{1,0,0},{2,0,0},{0,1,0},{1,1,0}},
            {{0,0,0},{1,0,0},{1,1,0},{2,1,0}}
        };
        for (int[][] s : shapeData) {
            List<int[]> list = new ArrayList<>();
            for (int[] p : s) list.add(p.clone());
            shapes.add(list);
        }
        nextPiece = shapes.get(rand.nextInt(shapes.size()));
        spawnPiece();
    }

    private boolean isValid(List<int[]> piece, int[] pos) {
        int px=pos[0], py=pos[1], pz=pos[2];
        for (int[] p : piece) {
            int x=px+p[0], y=py+p[1], z=pz+p[2];
            if (x<0||x>=W||y<0||y>=H||z<0||z>=D) return false;
            if (field[x][y][z]!=0) return false;
        }
        return true;
    }

    private void placePiece() {
        int px=currentPos[0], py=currentPos[1], pz=currentPos[2];
        for (int[] p : currentPiece) {
            field[px+p[0]][py+p[1]][pz+p[2]] = 1;
        }
        clearLayers();
        spawnPiece();
    }

    private void clearLayers() {
        int cleared = 0;
        for (int z=0; z<D; z++) {
            boolean full = true;
            for (int x=0; x<W; x++) for (int y=0; y<H; y++) if (field[x][y][z]==0) full=false;
            if (full) {
                for (int zz=z; zz<D-1; zz++) {
                    for (int x=0; x<W; x++) for (int y=0; y<H; y++) field[x][y][zz] = field[x][y][zz+1];
                }
                for (int x=0; x<W; x++) for (int y=0; y<H; y++) field[x][y][D-1] = 0;
                cleared++; z--;
            }
        }
        if (cleared>0) {
            score += cleared*100;
            level = 1 + score/500;
            fallInterval = Math.max(0.2, 1.0/(1+(level-1)*0.2));
        }
    }

    private void spawnPiece() {
        currentPiece = nextPiece;
        nextPiece = shapes.get(rand.nextInt(shapes.size()));
        currentPos = new int[]{1,1,3};
        if (!isValid(currentPiece, currentPos)) gameOver = true;
    }

    private boolean move(int dx, int dy, int dz) {
        int[] newPos = {currentPos[0]+dx, currentPos[1]+dy, currentPos[2]+dz};
        if (isValid(currentPiece, newPos)) {
            currentPos = newPos;
            return true;
        }
        return false;
    }

    private void rotate(String axis) {
        List<int[]> newPiece = new ArrayList<>();
        if (axis.equals("z")) {
            for (int[] p : currentPiece) newPiece.add(new int[]{-p[1], p[0], p[2]});
        } else if (axis.equals("x")) {
            for (int[] p : currentPiece) newPiece.add(new int[]{p[0], -p[2], p[1]});
        } else if (axis.equals("y")) {
            for (int[] p : currentPiece) newPiece.add(new int[]{p[2], p[1], -p[0]});
        }
        if (isValid(newPiece, currentPos)) currentPiece = newPiece;
    }

    private void hardDrop() {
        while (isValid(currentPiece, new int[]{currentPos[0], currentPos[1], currentPos[2]-1})) {
            currentPos[2]--;
        }
        placePiece();
    }

    private void update() {
        if (isValid(currentPiece, new int[]{currentPos[0], currentPos[1], currentPos[2]-1})) {
            currentPos[2]--;
        } else {
            placePiece();
        }
    }

    private void draw() {
        System.out.print("\033[H\033[2J");
        System.out.flush();
        System.out.println("ТЕТРИС 3D");
        System.out.println("Очки: "+score+" | Уровень: "+level);
        for (int z=D-1; z>=0; z--) {
            System.out.println("\nСЛОЙ "+(z+1)+" (Y="+z+")");
            System.out.println("  0 1 2 3");
            for (int y=H-1; y>=0; y--) {
                System.out.print(y+" ");
                for (int x=0; x<W; x++) {
                    if (field[x][y][z]!=0) System.out.print("X ");
                    else {
                        boolean inPiece = false;
                        for (int[] p : currentPiece) {
                            if (x==currentPos[0]+p[0] && y==currentPos[1]+p[1] && z==currentPos[2]+p[2]) inPiece=true;
                        }
                        System.out.print(inPiece ? "█ " : ". ");
                    }
                }
                System.out.println();
            }
        }
        System.out.println("\nСледующая фигура:");
        for (int dy=0; dy<2; dy++) {
            for (int dx=0; dx<2; dx++) {
                boolean found = false;
                for (int[] p : nextPiece) if (p[0]==dx && p[1]==dy && p[2]==0) found=true;
                System.out.print(found ? "█ " : ". ");
            }
            System.out.println();
        }
        System.out.println("\nWASD - движение, Q/E - поворот Z, R - поворот X, Space - падение, Esc - выход");
    }

    public void run() throws IOException {
        long lastFall = System.currentTimeMillis();
        while (!gameOver) {
            draw();
            // чтение ввода (неблокирующее с использованием System.in.available())
            if (System.in.available() > 0) {
                char ch = (char)System.in.read();
                if (ch == 'a') move(-1,0,0);
                else if (ch == 'd') move(1,0,0);
                else if (ch == 'w') move(0,1,0);
                else if (ch == 's') move(0,-1,0);
                else if (ch == 'q') rotate("z");
                else if (ch == 'e') { rotate("z"); rotate("z"); rotate("z"); }
                else if (ch == 'r') rotate("x");
                else if (ch == ' ') hardDrop();
                else if (ch == 27) break; // Esc
            }
            long now = System.currentTimeMillis();
            if ((now - lastFall) > fallInterval*1000) {
                update();
                lastFall = now;
            }
            try { Thread.sleep(50); } catch (InterruptedException e) {}
        }
        System.out.println("ИГРА ОКОНЧЕНА! Счёт: "+score);
    }

    public static void main(String[] args) throws IOException {
        new Tetris3D().run();
    }
}

File: tetris3d.rs
// tetris3d.rs — трёхмерный тетрис на Rust

use std::io::{self, Write, stdin, stdout};
use std::thread;
use std::time::{Duration, Instant};
use std::collections::VecDeque;
use rand::Rng;
use termion::{clear, cursor, color, style};
use termion::input::TermRead;

const W: usize = 4;
const H: usize = 4;
const D: usize = 4;

type Point = (i32, i32, i32);
type Piece = Vec<Point>;

struct Tetris3D {
    field: [[[u8; D]; H]; W],
    score: i32,
    level: i32,
    fall_interval: f64,
    current_piece: Piece,
    next_piece: Piece,
    current_pos: Point,
    game_over: bool,
}

impl Tetris3D {
    fn new() -> Self {
        let shapes: Vec<Piece> = vec![
            vec![(0,0,0),(1,0,0),(0,1,0),(1,1,0)],
            vec![(0,0,0),(1,0,0),(2,0,0),(3,0,0)],
            vec![(0,0,0),(1,0,0),(2,0,0),(1,1,0)],
            vec![(0,0,0),(1,0,0),(2,0,0),(0,1,0)],
            vec![(0,0,0),(1,0,0),(2,0,0),(2,1,0)],
            vec![(1,0,0),(2,0,0),(0,1,0),(1,1,0)],
            vec![(0,0,0),(1,0,0),(1,1,0),(2,1,0)],
        ];
        let mut rng = rand::thread_rng();
        let next = shapes[rng.gen_range(0..shapes.len())].clone();
        let mut game = Tetris3D {
            field: [[[0; D]; H]; W],
            score: 0,
            level: 1,
            fall_interval: 1.0,
            current_piece: Vec::new(),
            next_piece: next,
            current_pos: (0,0,0),
            game_over: false,
        };
        game.spawn_piece();
        game
    }

    fn is_valid(&self, piece: &Piece, pos: Point) -> bool {
        let (px, py, pz) = pos;
        for &(dx, dy, dz) in piece {
            let x = px + dx;
            let y = py + dy;
            let z = pz + dz;
            if x < 0 || x >= W as i32 || y < 0 || y >= H as i32 || z < 0 || z >= D as i32 {
                return false;
            }
            if self.field[x as usize][y as usize][z as usize] != 0 {
                return false;
            }
        }
        true
    }

    fn place_piece(&mut self) {
        let (px, py, pz) = self.current_pos;
        for &(dx, dy, dz) in &self.current_piece {
            self.field[(px+dx) as usize][(py+dy) as usize][(pz+dz) as usize] = 1;
        }
        self.clear_layers();
        self.spawn_piece();
    }

    fn clear_layers(&mut self) {
        let mut cleared = 0;
        let mut z = 0;
        while z < D {
            let mut full = true;
            for x in 0..W {
                for y in 0..H {
                    if self.field[x][y][z] == 0 {
                        full = false;
                        break;
                    }
                }
                if !full { break; }
            }
            if full {
                for zz in z..D-1 {
                    for x in 0..W {
                        for y in 0..H {
                            self.field[x][y][zz] = self.field[x][y][zz+1];
                        }
                    }
                }
                for x in 0..W {
                    for y in 0..H {
                        self.field[x][y][D-1] = 0;
                    }
                }
                cleared += 1;
                // после сдвига, проверяем этот же уровень снова
            } else {
                z += 1;
            }
        }
        if cleared > 0 {
            self.score += cleared * 100;
            self.level = 1 + self.score / 500;
            self.fall_interval = (1.0 / (1.0 + (self.level-1) as f64 * 0.2)).max(0.2);
        }
    }

    fn spawn_piece(&mut self) {
        let shapes: Vec<Piece> = vec![
            vec![(0,0,0),(1,0,0),(0,1,0),(1,1,0)],
            vec![(0,0,0),(1,0,0),(2,0,0),(3,0,0)],
            vec![(0,0,0),(1,0,0),(2,0,0),(1,1,0)],
            vec![(0,0,0),(1,0,0),(2,0,0),(0,1,0)],
            vec![(0,0,0),(1,0,0),(2,0,0),(2,1,0)],
            vec![(1,0,0),(2,0,0),(0,1,0),(1,1,0)],
            vec![(0,0,0),(1,0,0),(1,1,0),(2,1,0)],
        ];
        let mut rng = rand::thread_rng();
        self.current_piece = self.next_piece.clone();
        self.next_piece = shapes[rng.gen_range(0..shapes.len())].clone();
        self.current_pos = (1,1,3);
        if !self.is_valid(&self.current_piece, self.current_pos) {
            self.game_over = true;
        }
    }

    fn move_piece(&mut self, dx: i32, dy: i32, dz: i32) {
        let new_pos = (self.current_pos.0 + dx, self.current_pos.1 + dy, self.current_pos.2 + dz);
        if self.is_valid(&self.current_piece, new_pos) {
            self.current_pos = new_pos;
        }
    }

    fn rotate(&mut self, axis: char) {
        let new_piece: Piece = self.current_piece.iter().map(|&(x,y,z)| {
            match axis {
                'z' => (-y, x, z),
                'x' => (x, -z, y),
                'y' => (z, y, -x),
                _ => (x,y,z),
            }
        }).collect();
        if self.is_valid(&new_piece, self.current_pos) {
            self.current_piece = new_piece;
        }
    }

    fn hard_drop(&mut self) {
        while self.is_valid(&self.current_piece, (self.current_pos.0, self.current_pos.1, self.current_pos.2 - 1)) {
            self.current_pos.2 -= 1;
        }
        self.place_piece();
    }

    fn update(&mut self) {
        if self.is_valid(&self.current_piece, (self.current_pos.0, self.current_pos.1, self.current_pos.2 - 1)) {
            self.current_pos.2 -= 1;
        } else {
            self.place_piece();
        }
    }

    fn draw(&self) {
        print!("{}{}", clear::All, cursor::Goto(1,1));
        println!("ТЕТРИС 3D");
        println!("Очки: {} | Уровень: {}", self.score, self.level);
        for z in (0..D).rev() {
            println!("\nСЛОЙ {} (Y={})", z+1, z);
            println!("  0 1 2 3");
            for y in (0..H).rev() {
                print!("{} ", y);
                for x in 0..W {
                    if self.field[x][y][z] != 0 {
                        print!("X ");
                    } else {
                        let mut in_piece = false;
                        for &(dx,dy,dz) in &self.current_piece {
                            if x as i32 == self.current_pos.0+dx && y as i32 == self.current_pos.1+dy && z as i32 == self.current_pos.2+dz {
                                in_piece = true;
                                break;
                            }
                        }
                        print!("{} ", if in_piece { "█" } else { "." });
                    }
                }
                println!();
            }
        }
        println!("\nСледующая фигура:");
        for dy in 0..2 {
            for dx in 0..2 {
                let mut found = false;
                for &(ex,ey,ez) in &self.next_piece {
                    if ex == dx && ey == dy && ez == 0 {
                        found = true;
                        break;
                    }
                }
                print!("{} ", if found { "█" } else { "." });
            }
            println!();
        }
        println!("\nWASD - движение, Q/E - поворот Z, R - поворот X, Space - падение, Esc - выход");
        stdout().flush().unwrap();
    }

    fn run(&mut self) {
        let mut last_fall = Instant::now();
        let stdin = stdin();
        let mut keys = stdin.keys();

        while !self.game_over {
            self.draw();
            // Проверка ввода
            if let Some(Ok(key)) = keys.next() {
                match key {
                    termion::event::Key::Char('a') => self.move_piece(-1,0,0),
                    termion::event::Key::Char('d') => self.move_piece(1,0,0),
                    termion::event::Key::Char('w') => self.move_piece(0,1,0),
                    termion::event::Key::Char('s') => self.move_piece(0,-1,0),
                    termion::event::Key::Char('q') => self.rotate('z'),
                    termion::event::Key::Char('e') => { self.rotate('z'); self.rotate('z'); self.rotate('z'); },
                    termion::event::Key::Char('r') => self.rotate('x'),
                    termion::event::Key::Char(' ') => self.hard_drop(),
                    termion::event::Key::Esc => break,
                    _ => {}
                }
            }
            if last_fall.elapsed().as_secs_f64() > self.fall_interval {
                self.update();
                last_fall = Instant::now();
            }
            thread::sleep(Duration::from_millis(50));
        }
        println!("ИГРА ОКОНЧЕНА! Счёт: {}", self.score);
    }
}

fn main() {
    let mut game = Tetris3D::new();
    game.run();
}
